src/d3d9/d3d9_device.cpp

```diff
@@ -6972,6 +6981,68 @@ namespace dxvk {
   }
 
 
+  void D3D9DeviceEx::EnsureSamplerLimit() {
+    constexpr uint32_t MaxSamplerCount = DxvkSamplerPool::MaxSamplerCount - SamplerCount;
+
+    // Maximum possible number of live samplers we can have
+    // since last reading back from the CS thread.
+    if (likely(m_lastSamplerLiveCount + m_samplerBindCount - m_lastSamplerBindCount <= MaxSamplerCount))
+      return;
+
+    // Update current stats from CS thread and check again. We
+    // don't want to do this every time due to potential cache
+    // thrashing.
+    uint64_t lastStats = m_lastSamplerStats.load(std::memory_order_relaxed);
+    m_lastSamplerLiveCount = lastStats & SamplerCountMask;
+    m_lastSamplerBindCount = lastStats >> SamplerCountBits;
+
+    if (likely(m_lastSamplerLiveCount + m_samplerBindCount - m_lastSamplerBindCount <= MaxSamplerCount))
+      return;
+
+    // If we have a large number of sampler updates in flight, wait for
+    // the CS thread to complete some and re-evaluate. We should not hit
+    // this path under normal gameplay conditions.
+    ConsiderFlush(GpuFlushType::ImplicitSynchronization);
+
+    uint64_t sequenceNumber = m_csThread.lastSequenceNumber();
+
+    while (++sequenceNumber <= GetCurrentSequenceNumber()) {
+      SynchronizeCsThread(sequenceNumber);
+
+      uint64_t lastStats = m_lastSamplerStats.load(std::memory_order_relaxed);
+      m_lastSamplerLiveCount = lastStats & SamplerCountMask;
+      m_lastSamplerBindCount = lastStats >> SamplerCountBits;
+
+      if (m_lastSamplerLiveCount + m_samplerBindCount - m_lastSamplerBindCount <= MaxSamplerCount)
+        return;
+    }
+
+    // If we end up here, the game somehow managed to queue up so
+    // many samplers that we need to wait for the GPU to free some.
+    // We should absolutely never hit this path in the real world.
+    Logger::warn("Sampler pool exhausted, synchronizing with GPU.");
+
+    Flush();
+    SynchronizeCsThread(DxvkCsThread::SynchronizeAll);
+
+    // Need to reset sampler stats here since we'd
+    // otherwise hit this path again the next time
+    m_lastSamplerStats = 0u;
```

Comment:
Shouldn't this be `m_lastSamplerStats = m_samplerBindCount << SamplerCountBits;`? Otherwise, the next time `m_samplerBindCount` is still some high value, and m_lastSamplerBindCount won't offset it because it's 0.

Reply:
Or maybe even move it just before the return and store `(m_samplerBindCount << SamplerCountBits) | m_dxvkDevice->getSamplerStats().liveCount`. Or maybe I'm too stupid to understand threaded code, that's also always a possibility. :frog:

Reply:
Yeah, I think we do actually need to update this with the live count while CS is idle.

The idea here was that CS would give us new stats before the next call, but that's not guaranteed to happen.

Reply:
Fixed.